torch/csrc/Stream.cpp

```diff
@@ -256,6 +256,65 @@ static PyObject* THPStream_eq(THPStream* self, THPStream* other) {
   END_HANDLE_TH_ERRORS
 }
 
+static PyObject* THPStream_enter(THPStream* self, PyObject* unused) {
+  HANDLE_TH_ERRORS
+  // No operation is performed if the stream does not belong to an accelerator.
+  c10::DeviceType dst_device_type =
+      static_cast<c10::DeviceType>(self->device_type);
+  if (C10_UNLIKELY(!at::accelerator::isAccelerator(dst_device_type))) {
+    Py_RETURN_NONE;
+  }
+  c10::DeviceIndex src_device_idx = at::accelerator::getDeviceIndex();
+  c10::DeviceIndex dst_device_idx =
+      static_cast<c10::DeviceIndex>(self->device_index);
+  // If the stream is not on the current device, switch the current device to
+  // the device of the stream.
+  if (dst_device_idx != src_device_idx) {
+    at::accelerator::setDeviceIndex(dst_device_idx);
+  }
+  c10::Stream dst_prev_stream =
+      at::accelerator::getCurrentStream(dst_device_idx);
+  at::accelerator::setCurrentStream(
+      c10::Stream::unpack3(self->stream_id, dst_device_idx, dst_device_type));
+  auto src_device_index_object = THPUtils_packDeviceIndex(src_device_idx);
+  auto dst_prev_stream_object = THPStream_Wrap(dst_prev_stream);
+  // Push [src, dst] streams onto the stack.
+  at::impl::PythonTorchFunctionTLS::push_onto_stack(
+      std::make_shared<c10::SafePyObject>(
+          src_device_index_object, getPyInterpreter()));
+  at::impl::PythonTorchFunctionTLS::push_onto_stack(
+      std::make_shared<c10::SafePyObject>(
+          dst_prev_stream_object, getPyInterpreter()));
```

Comment:
lol, where did you found this pattern being used?

This is absolutely not for this use. If you do any op within the context manager, I expect it is going to blow up in a pretty fun way :D  Your test just happens not to run any op and so it happens to work.

The way you have to do this in the raw C API is to add two fields onto the struct for these two things if you want them to be always there.
If you want them to be added dynamically, you will have to enable dict attribute on this object which I don't recall exactly what's the exact way to do it.

I will repeat what I said above: writing it in python would be simpler haha

Reply:
I learned this pattern from https://github.com/pytorch/pytorch/blob/e4ecb09b3513e0ee53ed87496d8bfdf5d2944042/torch/csrc/Device.cpp#L179 Previously, I assume that pushing and popping on the stack always needed to be paired. However, I ended up using push and pop to implement the context manager here, which feels unintuitive. I don't think adding two additional fields to the Stream struct makes much sense either.

I also think it would be easier to write it in Python. I consider two options for the design and am unsure which one is better:
1. Separate the functionality of streams and context managers: We could introduce a `torch.accelerator.stream` function or `torch.accelerator.StreamGuard` class specifically for context management. This approach, previously mentioned in our [discussion](https://github.com/pytorch/pytorch/pull/132204#discussion_r1787965005),  avoids modifying the Stream struct.
2. Introduce `torch._C._StreamBase` class: We could change the `torch.Stream`, an immutable class that can't be modified in Python, in https://github.com/pytorch/pytorch/blob/e3886fb13c0719044da32571510ed7e3b9d75a94/torch/csrc/Stream.cpp#L328 into a base C++ class, `torch._C._StreamBase`. Then, in Python, create a new `torch.Stream` class derived from `torch._C._StreamBase`(e.g. in torch/streams.py). This new class would implement `__enter__` and `__exit__` methods for context manager.

I would love feedback on these two approaches.